src/filesystem/impls/filer/UrlCache.js

```diff
@@ -0,0 +1,381 @@
+        var self = this;
+        var removed = [];
+
+        // If this is a dir path, look for other paths entries below it
+        Object.keys(self.urls).forEach(function(key) {
+            var url = self.urls[key];
+
+            // The first time a file is written, we won't have
+            // a stale cache entry to clean up.
+            if(!url) {
+                return;
+            }
+
+            // If this filename matches exactly, or is a root path (i.e., other
+            // filenames begin with "<path>/...", remove it. Otherwise just skip.
+            if(key === path || key.indexOf(path + "/") === 0) {
+                removed.push(key);
+
+                delete self.urls[key];
+                delete self.paths[url];
+
+                // Delete the reference from memory
+                URL.revokeObjectURL(url);
+            }
+        });
+
+        _.defer(callback, null, removed);
+    };
+    BlobUrlProvider.prototype.rename = function(oldPath, newPath, callback) {
+        var url = this.urls[oldPath];
+
+        this.urls[newPath] = url;
+        this.paths[url] = newPath;
+        delete this.urls[oldPath];
+
+        _.defer(callback);
+    };
+
+
+    /**
+     * CacheStorageUrlProvider uses CacheStorage and Service Workers in compatible browsers.
+     */
+    function CacheStorageUrlProvider() {
+        UrlProviderBase.call(this);
+
+        // We use cache URLs like https://<origin>/dist/vfs/project/root/filename.ext
+        var prefix = brackets.env === "production" ? "/dist" : "/src";
+        this.generateVFSUrlForPath = function(path) {
+            var a = document.createElement("a");
+            a.href = prefix + "/vfs" + path;
+            return a.href;
+        };
+
+        this.projectCacheName = Path.join("vfs", StartupState.project("root"));
+        this.baseUrl = this.generateVFSUrlForPath(StartupState.project("root")) + "/";
+        this.shouldRewriteUrls = false;
+    }
+    CacheStorageUrlProvider.prototype = Object.create(UrlProviderBase.prototype);
+    CacheStorageUrlProvider.prototype.constructor = CacheStorageUrlProvider;
+
+    CacheStorageUrlProvider.prototype.clear = function(callback) {
+        var self = this;
+        var projectCacheName = self.projectCacheName;
+
+        // Delete existing cache for this root, and recreate empty cache.
+        window.caches
+            .delete(projectCacheName)
+            .then(function() {
+                caches.open(projectCacheName).then(function() {
+                    self.paths = {};
+                    self.urls = {};
+                    callback();
+                });
+            })
+            .catch(callback);
+    };
+    CacheStorageUrlProvider.prototype.init = function(callback) {
+        this.clear(callback);
+    };
+    CacheStorageUrlProvider.prototype.createURL = function(filename, blob, type, callback) {
+        var response = new Response(blob, {
```

Comment:
I'm curious about how this is passing linting since it isn't defined in the `global` comment at the beginning of the file